### PR TITLE
[WIP] Functional composition

### PR DESCRIPTION
**Pull request by alexeyshockov**
Some useful functions for functional composition.

They are without tests and documentation now, because I want to discuss the topic first.

### CODE REVIEW

File: src/Functional/Compose.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * Composes two functions together, so result function will be f(g(...))
+ *
+ * @param callable $f
+ * @param callable $g
+ *
+ * @return \Closure
+ */
+function compose(callable $f, callable $g)
+{
+    return function (...$args) use ($f, $g) {
```
Comment (lstrojny) [inline src/Functional/Compose.php]: Couldn’t that be an alias to `and_then()`

File: src/Functional/AndThen.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Combines two functions together (one after another), so result function will be g(f(...))
+ *
+ * @param callable $f
+ * @param callable $g
+ *
+ * @return \Closure
+ */
+function and_then(callable $f, callable $g)
```
Comment (lstrojny) [inline src/Functional/AndThen.php]: Wouldn’t we want to compose n functions by that? So `and_then(fn1, fn2, fn3, ...)`

File: src/Functional/OrElse.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Produces a new function, where the original function f() is replaced (in case of exception) by another function g()
+ *
+ * @param callable $f
+ * @param callable $g
+ *
+ * @return \Closure
+ */
+function or_else(callable $f, callable $g)
```
Comment (lstrojny) [inline src/Functional/OrElse.php]: Same here, I would expect to compose n functions

Comment (lstrojny) [inline src/Functional/OrElse.php]: Ah, wait, that’s actually not going to work :)

**Comment by lstrojny** [comment]
I generally like the idea of including functional composition capabilities

File: src/Functional/Compose.php (hunk 1 above)
Comment (alexeyshockov) [inline src/Functional/Compose.php]: I think reverse alias is better (`and_then()` is an alias for `compose()`). Done this.

File: src/Functional/AndThen.php (hunk 2 above)
Comment (alexeyshockov) [inline src/Functional/AndThen.php]: Yes, we would want. Will do this.

File: src/Functional/OrElse.php (hunk 3 above)
Comment (alexeyshockov) [inline src/Functional/OrElse.php]: Why? I wrote some tests, they pass.

Comment (lstrojny) [inline src/Functional/OrElse.php]: No, my idea doesn’t work, not your code. Your code is fine :+1:

File: src/Functional/Compose.php (hunk 1 above)
Comment (lstrojny) [inline src/Functional/Compose.php]: Makes sense

File: src/Functional/IfNotDefined.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+ */
+namespace Functional;
+
+/**
+ * @param callable $f
+ * @param callable $g
+ *
+ * @return \Closure
+ */
+function if_not_defined(callable $f, callable $g)
```
Comment (lstrojny) [inline src/Functional/IfNotDefined.php]: Hmm, what does that do?

Comment (alexeyshockov) [inline src/Functional/IfNotDefined.php]: Sorry, this function is experimental and present here by my mistake. I will remove it :)

BTW, idea is the same as in the Scala's PartialFunction.orElse (http://www.scala-lang.org/api/current/index.html#scala.PartialFunction).

**Comment by mschwager** [comment]
I've also thought of contributing a `compose` function, but why limit it to two functions? Here's what I came up with:

```php
function identity($x) {
    return $x;
}

function compose(...$functions) {
    return array_reduce(
        $functions,
        function ($carry, $item) {
            return function ($x) use ($carry, $item) {
                return $item($carry($x));
            };
        },
        "identity"
    );
}

$plus2 = function ($x) { return $x + 2; };
$times4 = function ($x) { return $x * 4; };
$square = function ($x) { return $x * $x; };

$composed = compose($plus2, $times4, $square);

$mapped = array_map($composed, array(1,2,8,5));

assert($mapped == array(144, 256, 1600, 784));
```

**Comment by lstrojny** [comment]
@alexeyshockov I like @mschwager’s idea about compose() being n-ary

**Comment by lstrojny** [comment]
Closing, as we have a `compose()` now.

**Pull request closed without merging**